// firebase-tools/lib/api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.githubOrigin = exports.apphostingOrigin = exports.serviceUsageOrigin = exports.cloudRunApiOrigin = exports.hostingApiOrigin = exports.firebaseStorageOrigin = exports.storageOrigin = exports.runtimeconfigOrigin = exports.rulesOrigin = exports.resourceManagerOrigin = exports.remoteConfigApiOrigin = exports.rtdbMetadataOrigin = exports.rtdbManagementOrigin = exports.realtimeOrigin = exports.extensionsTOSOrigin = exports.extensionsPublisherOrigin = exports.extensionsOrigin = exports.iamOrigin = exports.identityOrigin = exports.hostingOrigin = exports.googleOrigin = exports.pubsubOrigin = exports.cloudTasksOrigin = exports.cloudschedulerOrigin = exports.developerConnectP4SAOrigin = exports.developerConnectOrigin = exports.cloudbuildOrigin = exports.functionsDefaultRegion = exports.runOrigin = exports.functionsV2Origin = exports.functionsOrigin = exports.firestoreOrigin = exports.firestoreOriginOrEmulator = exports.firedataOrigin = exports.firebaseExtensionsRegistryOrigin = exports.firebaseApiOrigin = exports.eventarcOrigin = exports.dynamicLinksKey = exports.dynamicLinksOrigin = exports.consoleOrigin = exports.authOrigin = exports.appDistributionOrigin = exports.artifactRegistryDomain = exports.containerRegistryDomain = exports.cloudMonitoringOrigin = exports.cloudloggingOrigin = exports.cloudbillingOrigin = exports.clientSecret = exports.clientId = exports.authProxyOrigin = void 0;
exports.setScopes = exports.getScopes = exports.githubClientSecret = exports.githubClientId = exports.computeOrigin = exports.secretManagerOrigin = exports.githubApiOrigin = void 0;
const constants_1 = require("./emulator/constants");
const logger_1 = require("./logger");
const scopes = require("./scopes");
const utils = require("./utils");
let commandScopes = new Set();
const authProxyOrigin = () => utils.envOverride("FIREBASE_AUTHPROXY_URL", "https://auth.firebase.tools");
exports.authProxyOrigin = authProxyOrigin;
const clientId = () => utils.envOverride("FIREBASE_CLIENT_ID", "563584335869-fgrhgmd47bqnekij5i8b5pr03ho849e6.apps.googleusercontent.com");
exports.clientId = clientId;
const clientSecret = () => utils.envOverride("FIREBASE_CLIENT_SECRET", "j9iVZfS8kkCEFUPaAeJV0sAi");
exports.clientSecret = clientSecret;
const cloudbillingOrigin = () => utils.envOverride("FIREBASE_CLOUDBILLING_URL", "https://cloudbilling.googleapis.com");
exports.cloudbillingOrigin = cloudbillingOrigin;
const cloudloggingOrigin = () => utils.envOverride("FIREBASE_CLOUDLOGGING_URL", "https://logging.googleapis.com");
exports.cloudloggingOrigin = cloudloggingOrigin;
const cloudMonitoringOrigin = () => utils.envOverride("CLOUD_MONITORING_URL", "https://monitoring.googleapis.com");
exports.cloudMonitoringOrigin = cloudMonitoringOrigin;
const containerRegistryDomain = () => utils.envOverride("CONTAINER_REGISTRY_DOMAIN", "gcr.io");
exports.containerRegistryDomain = containerRegistryDomain;
const artifactRegistryDomain = () => utils.envOverride("ARTIFACT_REGISTRY_DOMAIN", "https://artifactregistry.googleapis.com");
exports.artifactRegistryDomain = artifactRegistryDomain;
const appDistributionOrigin = () => utils.envOverride("FIREBASE_APP_DISTRIBUTION_URL", "https://firebaseappdistribution.googleapis.com");
exports.appDistributionOrigin = appDistributionOrigin;
const authOrigin = () => utils.envOverride("FIREBASE_AUTH_URL", "https://accounts.google.com");
exports.authOrigin = authOrigin;
const consoleOrigin = () => utils.envOverride("FIREBASE_CONSOLE_URL", "https://console.firebase.google.com");
exports.consoleOrigin = consoleOrigin;
const dynamicLinksOrigin = () => utils.envOverride("FIREBASE_DYNAMIC_LINKS_URL", "https://firebasedynamiclinks.googleapis.com");
exports.dynamicLinksOrigin = dynamicLinksOrigin;
const dynamicLinksKey = () => utils.envOverride("FIREBASE_DYNAMIC_LINKS_KEY", "AIzaSyB6PtY5vuiSB8MNgt20mQffkOlunZnHYiQ");
exports.dynamicLinksKey = dynamicLinksKey;
const eventarcOrigin = () => utils.envOverride("EVENTARC_URL", "https://eventarc.googleapis.com");
exports.eventarcOrigin = eventarcOrigin;
const firebaseApiOrigin = () => utils.envOverride("FIREBASE_API_URL", "https://firebase.googleapis.com");
exports.firebaseApiOrigin = firebaseApiOrigin;
const firebaseExtensionsRegistryOrigin = () => utils.envOverride("FIREBASE_EXT_REGISTRY_ORIGIN", "https://extensions-registry.firebaseapp.com");
exports.firebaseExtensionsRegistryOrigin = firebaseExtensionsRegistryOrigin;
const firedataOrigin = () => utils.envOverride("FIREBASE_FIREDATA_URL", "https://mobilesdk-pa.googleapis.com");
exports.firedataOrigin = firedataOrigin;
const firestoreOriginOrEmulator = () => utils.envOverride(constants_1.Constants.FIRESTORE_EMULATOR_HOST, utils.envOverride("FIRESTORE_URL", "https://firestore.googleapis.com"), (val) => {
    if (val.startsWith("http")) {
        return val;
    }
    return `http://${val}`;
});
exports.firestoreOriginOrEmulator = firestoreOriginOrEmulator;
const firestoreOrigin = () => utils.envOverride("FIRESTORE_URL", "https://firestore.googleapis.com");
exports.firestoreOrigin = firestoreOrigin;
const functionsOrigin = () => utils.envOverride("FIREBASE_FUNCTIONS_URL", "https://cloudfunctions.googleapis.com");
exports.functionsOrigin = functionsOrigin;
const functionsV2Origin = () => utils.envOverride("FIREBASE_FUNCTIONS_V2_URL", "https://cloudfunctions.googleapis.com");
exports.functionsV2Origin = functionsV2Origin;
const runOrigin = () => utils.envOverride("CLOUD_RUN_URL", "https://run.googleapis.com");
exports.runOrigin = runOrigin;
const functionsDefaultRegion = () => utils.envOverride("FIREBASE_FUNCTIONS_DEFAULT_REGION", "us-central1");
exports.functionsDefaultRegion = functionsDefaultRegion;
const cloudbuildOrigin = () => utils.envOverride("FIREBASE_CLOUDBUILD_URL", "https://cloudbuild.googleapis.com");
exports.cloudbuildOrigin = cloudbuildOrigin;
const developerConnectOrigin = () => utils.envOverride("FIREBASE_DEVELOPERCONNECT_URL", "https://developerconnect.googleapis.com");
exports.developerConnectOrigin = developerConnectOrigin;
const developerConnectP4SAOrigin = () => utils.envOverride("FIREBASE_DEVELOPERCONNECT_P4SA_URL", "gcp-sa-devconnect.iam.gserviceaccount.com");
exports.developerConnectP4SAOrigin = developerConnectP4SAOrigin;
const cloudschedulerOrigin = () => utils.envOverride("FIREBASE_CLOUDSCHEDULER_URL", "https://cloudscheduler.googleapis.com");
exports.cloudschedulerOrigin = cloudschedulerOrigin;
const cloudTasksOrigin = () => utils.envOverride("FIREBASE_CLOUD_TAKS_URL", "https://cloudtasks.googleapis.com");
exports.cloudTasksOrigin = cloudTasksOrigin;
const pubsubOrigin = () => utils.envOverride("FIREBASE_PUBSUB_URL", "https://pubsub.googleapis.com");
exports.pubsubOrigin = pubsubOrigin;
const googleOrigin = () => utils.envOverride("FIREBASE_TOKEN_URL", utils.envOverride("FIREBASE_GOOGLE_URL", "https://www.googleapis.com"));
exports.googleOrigin = googleOrigin;
const hostingOrigin = () => utils.envOverride("FIREBASE_HOSTING_URL", "https://web.app");
exports.hostingOrigin = hostingOrigin;
const identityOrigin = () => utils.envOverride("FIREBASE_IDENTITY_URL", "https://identitytoolkit.googleapis.com");
exports.identityOrigin = identityOrigin;
const iamOrigin = () => utils.envOverride("FIREBASE_IAM_URL", "https://iam.googleapis.com");
exports.iamOrigin = iamOrigin;
const extensionsOrigin = () => utils.envOverride("FIREBASE_EXT_URL", "https://firebaseextensions.googleapis.com");
exports.extensionsOrigin = extensionsOrigin;
const extensionsPublisherOrigin = () => utils.envOverride("FIREBASE_EXT_PUBLISHER_URL", "https://firebaseextensionspublisher.googleapis.com");
exports.extensionsPublisherOrigin = extensionsPublisherOrigin;
const extensionsTOSOrigin = () => utils.envOverride("FIREBASE_EXT_TOS_URL", "https://firebaseextensionstos-pa.googleapis.com");
exports.extensionsTOSOrigin = extensionsTOSOrigin;
const realtimeOrigin = () => utils.envOverride("FIREBASE_REALTIME_URL", "https://firebaseio.com");
exports.realtimeOrigin = realtimeOrigin;
const rtdbManagementOrigin = () => utils.envOverride("FIREBASE_RTDB_MANAGEMENT_URL", "https://firebasedatabase.googleapis.com");
exports.rtdbManagementOrigin = rtdbManagementOrigin;
const rtdbMetadataOrigin = () => utils.envOverride("FIREBASE_RTDB_METADATA_URL", "https://metadata-dot-firebase-prod.appspot.com");
exports.rtdbMetadataOrigin = rtdbMetadataOrigin;
const remoteConfigApiOrigin = () => utils.envOverride("FIREBASE_REMOTE_CONFIG_URL", "https://firebaseremoteconfig.googleapis.com");
exports.remoteConfigApiOrigin = remoteConfigApiOrigin;
const resourceManagerOrigin = () => utils.envOverride("FIREBASE_RESOURCEMANAGER_URL", "https://cloudresourcemanager.googleapis.com");
exports.resourceManagerOrigin = resourceManagerOrigin;
const rulesOrigin = () => utils.envOverride("FIREBASE_RULES_URL", "https://firebaserules.googleapis.com");
exports.rulesOrigin = rulesOrigin;
const runtimeconfigOrigin = () => utils.envOverride("FIREBASE_RUNTIMECONFIG_URL", "https://runtimeconfig.googleapis.com");
exports.runtimeconfigOrigin = runtimeconfigOrigin;
const storageOrigin = () => utils.envOverride("FIREBASE_STORAGE_URL", "https://storage.googleapis.com");
exports.storageOrigin = storageOrigin;
const firebaseStorageOrigin = () => utils.envOverride("FIREBASE_FIREBASESTORAGE_URL", "https://firebasestorage.googleapis.com");
exports.firebaseStorageOrigin = firebaseStorageOrigin;
const hostingApiOrigin = () => utils.envOverride("FIREBASE_HOSTING_API_URL", "https://firebasehosting.googleapis.com");
exports.hostingApiOrigin = hostingApiOrigin;
const cloudRunApiOrigin = () => utils.envOverride("CLOUD_RUN_API_URL", "https://run.googleapis.com");
exports.cloudRunApiOrigin = cloudRunApiOrigin;
const serviceUsageOrigin = () => utils.envOverride("FIREBASE_SERVICE_USAGE_URL", "https://serviceusage.googleapis.com");
exports.serviceUsageOrigin = serviceUsageOrigin;
const apphostingOrigin = () => utils.envOverride("APPHOSTING_URL", "https://firebaseapphosting.googleapis.com");
exports.apphostingOrigin = apphostingOrigin;
const githubOrigin = () => utils.envOverride("GITHUB_URL", "https://github.com");
exports.githubOrigin = githubOrigin;
const githubApiOrigin = () => utils.envOverride("GITHUB_API_URL", "https://api.github.com");
exports.githubApiOrigin = githubApiOrigin;
const secretManagerOrigin = () => utils.envOverride("CLOUD_SECRET_MANAGER_URL", "https://secretmanager.googleapis.com");
exports.secretManagerOrigin = secretManagerOrigin;
const computeOrigin = () => utils.envOverride("COMPUTE_URL", "https://compute.googleapis.com");
exports.computeOrigin = computeOrigin;
const githubClientId = () => utils.envOverride("GITHUB_CLIENT_ID", "89cf50f02ac6aaed3484");
exports.githubClientId = githubClientId;
const githubClientSecret = () => utils.envOverride("GITHUB_CLIENT_SECRET", "3330d14abc895d9a74d5f17cd7a00711fa2c5bf0");
exports.githubClientSecret = githubClientSecret;
function getScopes() {
    return Array.from(commandScopes);
}
exports.getScopes = getScopes;
function setScopes(sps = []) {
    commandScopes = new Set([
        scopes.EMAIL,
        scopes.OPENID,
        scopes.CLOUD_PROJECTS_READONLY,
        scopes.FIREBASE_PLATFORM,
    ]);
    for (const s of sps) {
        commandScopes.add(s);
    }
    logger_1.logger.debug("> command requires scopes:", Array.from(commandScopes));
}
exports.setScopes = setScopes;
